net/host.go

```diff
@@ -79,25 +107,43 @@ func (h *Host) P2pHost() P2pHost {
 	return h.h
 }
 
-// SetHandler sets the Handler instance used by the host.
-func (h *Host) SetHandler(handler Handler) {
-	fn := func() bool {
-		return len(handler.GetOffers()) == 0
+func (h *Host) advertisedNamespaces() []string {
+	provides := []string{""}
+
+	if len(h.makerHandler.GetOffers()) > 0 {
+		provides = append(provides, string(coins.ProvidesXMR))
 	}
 
-	h.handler = handler
-	h.h.SetShouldAdvertiseFunc(fn)
+	if h.isRelayer {
+		provides = append(provides, relayerProvidesStr)
+	}
+
+	return provides
+}
+
+// SetHandlers sets the maker and taker instances used by the host.
+func (h *Host) SetHandlers(makerHandler MakerHandler, takerHandler TakerHandler) {
+	h.makerHandler = makerHandler
+	h.takerHandler = takerHandler
+	h.h.SetAdvertisedNamespacesFunc(h.advertisedNamespaces)
+
+	h.h.SetStreamHandler(queryProtocolID, h.handleQueryStream)
+	h.h.SetStreamHandler(relayProtocolID, h.handleRelayStream)
```

Comment:
```suggestion
if h.isRelayer {
	h.h.SetStreamHandler(relayProtocolID, h.handleRelayStream)
}
```

we should only handle incoming relay streams if we're a relayer

Reply:
We might need to remove it in the follow up PR, but I put it in as suggested.